### PR TITLE
Improve `bevy_winit` documentation

### PR DESCRIPTION
**Pull request by maniwani**
Redo of #7590 since I messed up my branch.

# Objective

- Revise docs.
- Refactor event loop code a little bit to make it easier to follow.

## Solution

- Do the above.

---

### Migration Guide

- `UpdateMode::Reactive { max_wait: .. }` -> `UpdateMode::Reactive { wait: .. }`
- `UpdateMode::ReactiveLowPower { max_wait: .. }` -> `UpdateMode::ReactiveLowPower { wait: .. }`

### CODE REVIEW

**Review by alice-i-cecile — COMMENTED** [review]
These changes look good to me and clean things up, but this needs to be rebased.

**Review by alice-i-cecile — APPROVED** [review]

**Review by IceSentry — COMMENTED** [review]
In general this looks good to me. I can't comment on the unsafe part though.

File: crates/bevy_winit/src/lib.rs [hunk 1]
```diff
@@ -244,100 +190,89 @@ struct InputEvents<'w> {
...
-    active: bool,
-    /// Tracks whether or not an event has occurred this frame that would trigger an update in low
-    /// power mode. Should be reset at the end of every frame.
-    low_power_event: bool,
-    /// Tracks whether the event loop was started this frame because of a redraw request.
-    redraw_request_sent: bool,
-    /// Tracks if the event loop was started this frame because of a [`ControlFlow::WaitUntil`]
-    /// timeout.
-    timeout_reached: bool,
+struct WinitAppRunnerState {
```
[the remark this answers is not in the document]
Comment (IceSentry) [reply crates/bevy_winit/src/lib.rs:203]: Ah, it's the timeout thing now. It should probably be mentioned in the docs that it's used for low_power

File: crates/bevy_winit/src/lib.rs [hunk 2]
```diff
@@ -244,100 +190,89 @@ struct InputEvents<'w> {
...
+    file_drag_and_drop: EventWriter<'w, FileDragAndDrop>,
 
-#[derive(SystemParam)]
-struct CursorEvents<'w> {
     cursor_moved: EventWriter<'w, CursorMoved>,
     cursor_entered: EventWriter<'w, CursorEntered>,
     cursor_left: EventWriter<'w, CursorLeft>,
+
+    // device events
+    mouse_motion: EventWriter<'w, MouseMotion>,
```
[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_winit/src/lib.rs:200]: It's a `DeviceEvent` as opposed to a `WindowEvent`.

File: crates/bevy_winit/src/lib.rs [hunk 3]
```diff
@@ -353,161 +288,126 @@ pub fn winit_runner(mut app: App) {
...
             }
-        }
 
-        if let Some(app_exit_events) = app.world.get_resource::<Events<AppExit>>() {
-            if app_exit_event_reader.iter(app_exit_events).last().is_some() {
-                *control_flow = ControlFlow::Exit;
-                return;
+            if let Some(app_exit_events) = app.world.get_resource::<Events<AppExit>>() {
+                if app_exit_event_reader.iter(app_exit_events).last().is_some() {
+                    return;
```
[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_winit/src/lib.rs:294]: Oh whoops, I had temporarily moved this block out of the loop and then put it back in.

File: crates/bevy_winit/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_winit/src/lib.rs:203]: Oh, in the docs of `WinitAppRunnerState`. It's not just a low-power thing, it's for both `Reactive` and `ReactiveLowPower`.

[the remark this answers is not in the document]
Comment (IceSentry) [reply crates/bevy_winit/src/lib.rs:203]: Right, I meant in the doc in this struct. Looking at it standalone it's not clear why there's a timeout but I guess it doesn't make sense to look at this struct on it's own

File: crates/bevy_winit/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (IceSentry) [reply crates/bevy_winit/src/lib.rs:200]: Ah, so I guess it's just winit that's making this distinction. Then nothing to do here I guess

File: crates/bevy_winit/src/lib.rs [hunk 4]
```diff
@@ -63,14 +61,16 @@ use crate::web_resize::{CanvasParentResizeEventChannel, CanvasParentResizePlugin
 #[cfg(target_os = "android")]
 pub static ANDROID_APP: std::sync::OnceLock<AndroidApp> = std::sync::OnceLock::new();
 
-/// A [`Plugin`] that utilizes [`winit`] for window creation and event loop management.
+/// Integrates [`winit`], extending an [`App`] with capabilities for managing windows and
```
[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_winit/src/lib.rs:64]: idk, I meant "extending" as an "adding to", but I'll change the wording to
```rust
/// A [`Plugin`] that integrates [`winit`], adding its window management and event loop
/// capabilities to an [`App`].
```

File: crates/bevy_winit/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_winit/src/lib.rs:203]: I've changed it to `wait_elapsed`. Although, maybe I should flip it around so the `UpdateMode` structs have `timeout`.

**Comment by maniwani** [comment]
> I can't comment on the unsafe part though.

@IceSentry, I changed it back to doing the `SystemState::get_mut` and passing those references into the function. I was doing the lifetime transmute in #9122 because it's essential with those changes, but it's not essential here.

**Review by IceSentry — APPROVED** [review]
Alright, nothing else jumps out at me so LGTM

**Comment by ickk** [comment]
I'm guessing the addition of the generic `T` to functions taking `EventLoop<()>` is to support #9122? 

Otherwise in this PR I don't understand its purpose, since all the functions of note here are private, and `WinitPlugin` only explicitly constructs an `EventLoop<()>`; it adds some cognitive load. But if you're going to add it anyway in that other PR then its fine here

**Comment by maniwani** [comment]
> I'm guessing the addition of the generic `T` to functions taking `EventLoop<()>` is #9122?

Ah, yes, that's correct. I'll change it back though. It's small enough, and it'll be justified then.

File: crates/bevy_winit/src/lib.rs [hunk 5]
```diff
@@ -655,132 +656,134 @@ pub fn winit_runner(mut app: App) {
...
                 #[cfg(target_os = "android")]
                 {
-                    // Bevy doesn't support suspend/resume so we just exit
-                    // and Android will restart the application on resume
-                    // TODO: Save save some state and load on resume
+                    // Android sending this event invalidates all render surfaces.
+                    // TODO
+                    // Upon resume, check if the new render surfaces are compatible with the
+                    // existing render device. If not (which should basically never happen),
+                    // and *then* try to rebuild the renderer.
```
Comment (ickk) [inline crates/bevy_winit/src/lib.rs]:
```suggestion
                    // *then* try to rebuild the renderer.
```

**Review by ickk — APPROVED** [review]

**Review by Selene-Amanita — COMMENTED** [review]
I'm not sure I understood everything but I finally reviewed it.

Sorry for the delay.

I'm mostly concerned about https://github.com/bevyengine/bevy/pull/7609#discussion_r1275383474.

File: crates/bevy_winit/src/winit_config.rs [hunk 6]
```diff
@@ -69,45 +79,43 @@ impl Default for WinitSettings {
...
-    /// ## Note
-    ///
-    /// Once the app has executed all bevy systems and reaches the end of the event loop, there is
-    /// no way to force the app to wake and update again, unless a `winit` event (such as user
-    /// input, or the window being resized) is received or the time limit is reached.
+    /// The [`App`](bevy_app::App) will update in response to the following, until an
+    /// [`AppExit`](bevy_app::AppExit) event appears:
+    /// - enough time has elapsed since the previous update
+    /// - a redraw is requested
+    /// - new window or device events have appeared
```
[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_winit/src/winit_config.rs:97]: It's `winit` own distinction. There are events that come from the windows ([`WindowEvent`][1]) and events that come from the input devices with no link to a particular window ([`DeviceEvent`][2]).

[1]: https://docs.rs/winit/latest/winit/event/enum.WindowEvent.html
[2]: https://docs.rs/winit/latest/winit/event/enum.DeviceEvent.html

File: crates/bevy_winit/src/lib.rs [hunk 7]
```diff
@@ -655,132 +656,136 @@ pub fn winit_runner(mut app: App) {
...
+                    #[cfg(not(target_arch = "wasm32"))]
+                    let (
+                        commands,
+                        mut windows,
+                        event_writer,
+                        winit_windows,
+                        adapters,
+                        handlers,
+                        accessibility_requested,
+                    ) = create_window_system_state.get_mut(&mut app.world);
```
[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_winit/src/lib.rs:754]: Whoops, that was just a mistake.

I'm pretty sure it won't duplicate windows though. `create_windows` skips the known `Entity` values.

https://github.com/bevyengine/bevy/blob/08ea1d18aed1a7996c9abeef566a6bcc5e023fc2/crates/bevy_winit/src/system.rs#L50-L52

File: crates/bevy_winit/src/winit_config.rs [hunk 8]
```diff
@@ -69,45 +79,43 @@ impl Default for WinitSettings {
...
-    /// maximum wait time has elapsed.
-    ///
-    /// ## Note
-    ///
-    /// Once the app has executed all bevy systems and reaches the end of the event loop, there is
-    /// no way to force the app to wake and update again, unless a `winit` event (such as user
-    /// input, or the window being resized) is received or the time limit is reached.
+    /// The [`App`](bevy_app::App) will update in response to the following, until an
+    /// [`AppExit`](bevy_app::AppExit) event appears:
+    /// - enough time has elapsed since the previous update
```
[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_winit/src/winit_config.rs:95]: Why link back to `Reactive` a second time here when `ReactiveLowPower`'s `wait` field already has its own documentation?

File: crates/bevy_winit/src/lib.rs [hunk 9]
```diff
@@ -244,79 +241,83 @@ struct InputEvents<'w> {
...
 /// The default [`App::runner`] for the [`WinitPlugin`] plugin.
 ///
-/// Overriding the app's [runner](bevy_app::App::runner) while using `WinitPlugin` will bypass the `EventLoop`.
+/// Overriding the app's [runner](bevy_app::App::runner) while using `WinitPlugin` will bypass the
+/// `EventLoop`.
 pub fn winit_runner(mut app: App) {
-    // We remove this so that we have ownership over it.
     let mut event_loop = app
         .world
         .remove_non_send_resource::<EventLoop<()>>()
```
[the remark this answers is not in the document]
Comment (maniwani) [reply crates/bevy_winit/src/lib.rs:294]: Right.

**Pull request merged**